components/ws-proxy/pkg/proxy/infoprovider.go

```diff
@@ -413,25 +432,69 @@ func (c *workspaceInfoCache) doInsert(info *WorkspaceInfo) {
 	}
 }
 
-func (c *workspaceInfoCache) Delete(workspaceID string) {
+func (c *workspaceInfoCache) Delete(workspaceID string, instanceID string) {
 	c.cond.L.Lock()
 	defer c.cond.L.Unlock()
 
-	info, present := c.infos[workspaceID]
-	if !present || info == nil {
+	infos, present := c.infos[workspaceID]
+	if !present || infos == nil {
 		return
 	}
-	delete(c.coordsByPublicPort, info.IDEPublicPort)
-	delete(c.infos, workspaceID)
+
+	// Keep only the active instances and public port(s) for the workspace id
+	activeInfos := []*WorkspaceInfo{}
+	activePublicPorts := make(map[string]interface{})
+	var instanceIDEPublicPort string
+	for _, info := range infos {
+		if info.InstanceID != instanceID {
+			activeInfos = append(activeInfos, info)
+			activePublicPorts[info.IDEPublicPort] = true
+		} else {
+			instanceIDEPublicPort = info.IDEPublicPort
+		}
+	}
+
+	if len(activeInfos) == 0 {
+		log.WithField("workspaceID", workspaceID).WithField("instanceID", instanceID).Debug("deleting all instances for workspace")
```

Comment:
I reckon the log statements in here would be rather verbose in prod

Reply:
Ditto on the testing - removed